Firestore/core/src/firebase/firestore/model/field_value.cc

```diff
@@ -197,21 +176,17 @@ FieldValue FieldValue::Delete(const FieldPath& field_path) const {
   const std::string& child_name = field_path.first_segment();
   const ObjectValue::Map& object_map = object_value_->internal_value;
   if (field_path.size() == 1) {
-    // TODO(zxu): Once immutable type is available, rewrite these.
-    ObjectValue::Map copy = CopyExcept(object_map, child_name);
-    return FieldValue::FromMap(std::move(copy));
+    return FieldValue::FromMap(object_map.erase(child_name));
   } else {
     const auto iter = object_map.find(child_name);
-    if (iter == object_map.end() || iter->second.type() != Type::Object) {
+    if (iter != object_map.end() && iter->second.type() == Type::Object) {
+      FieldValue new_child = iter->second.Delete(field_path.PopFirst());
+      return SetChild(child_name, new_child);
+    } else {
       // If the found value isn't an object, it cannot contain the remaining
       // segments of the path. We don't actually change a primitive value to
       // an object for a delete.
       return *this;
-    } else {
-      ObjectValue::Map copy = CopyExcept(object_map, child_name);
```

Comment:
Why is this case removed?

Reply:
It isn't. But I reversed the if condition, so this case is now in the 'if' portion, whereas the `return *this` is now in the else block.

So why'd I invert the if condition?  ... I don't remember. :/  (Looking at it now, I *think* it was to keep the iter closer to its usage?)

Reply:
Ah, sorry, you're right.